### PR TITLE
Add iwds07 protocol tests and code

### PR DESCRIPTION
**Pull request by 1technophile**
from these data:
https://forum.pilight.org/showthread.php?tid=2952

### CODE REVIEW

**Comment by CurlyMoo** [comment]
Why the need for a new PR?

File: libs/pilight/protocols/433.92/iwds07.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
+/*
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Can you change this in MPLv2?

File: libs/pilight/protocols/433.92/iwds07.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*
+	Copyright (C) 2017 CurlyMo & wo_rasp
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.h]: Likewise

File: tests/protocols/iwds07.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	},
+	{
+		"1239 413 413 1239 1239 413 1239 413 413 1239 1239 413 413 1239 413 1239 413 1239 1239 413 1239 413 1239 413 413 1239 413 1239 1239 413 1239 413 413 1239 413 1239 413 1239 413 1239 1239 413 1239 413 1239 413 413 1239 413 14042",
+		"{\"unit\":52781,\"battery\":1,\"state\":\"closed\"}",
+		NULL,
+		0
+	},
+	{
+		"1236 412 412 1236 1236 412 1236 412 412 1236 1236 412 412 1236 1236 412 1236 412 1236 412 412 1236 412 1236 412 1236 1236 412 1236 412 1236 412 412 1236 412 1236 412 1236 412 1236 1236 412 412 1236 1236 412 412 1236 412 14008",
+        "{\"unit\":58285,\"battery\":1,\"state\":\"opened\"}",
```
Comment (CurlyMoo) [inline tests/protocols/iwds07.h]: Indentation is off here.

File: tests/protocols/iwds07.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		NULL,
+		0
+	}
+	{
+		"1242 414 414 1242 1242 414 1242 414 414 1242 1242 414 414 1656 1242 414 1242 414 1242 414 414 1242 414 1242 414 1242 1242 414 1242 414 1242 414 414 1242 414 1242 414 1242 414 1242 1242 414 1242 414 1242 414 414 1242 414 14076",
+        "{\"unit\":58285,\"battery\":1,\"state\":\"closed\"}",
```
Comment (CurlyMoo) [inline tests/protocols/iwds07.h]: I miss one with battery = 0

**Comment by 1technophile** [comment]
> Why the need for a new PR?

I didn't know if you wanted one for each branch or no.

**Comment by CurlyMoo** [comment]
Why didn't you ask for before opening and closing unrepearable PR's?

File: libs/pilight/protocols/433.92/iwds07.h (hunk 2 above)
Comment (1technophile) [reply libs/pilight/protocols/433.92/iwds07.h]: done

File: tests/protocols/iwds07.h (hunk 3 above)
Comment (1technophile) [reply tests/protocols/iwds07.h]: done

File: libs/pilight/protocols/433.92/iwds07.c (hunk 1 above)
Comment (1technophile) [reply libs/pilight/protocols/433.92/iwds07.c]: done

File: tests/protocols/iwds07.h (hunk 4 above)
Comment (1technophile) [reply tests/protocols/iwds07.h]: I don't have examples with battery to 0 at this time

**Comment by 1technophile** [comment]
I should have said:
I thought you wanted one for each branch.

For my own guidance why do you consider PR as "unrepearable", is it an issue for you if I close it and redo it?

**Comment by CurlyMoo** [comment]
> I thought you wanted one for each branch.

Yes i did said that, but you already opened one for rewrite, than closed it, and removed the original branch it originated from.

**Comment by CurlyMoo** [comment]
Can you sync with my latest commit so the unit test works again?

File: tests/protocols/iwds07.h (hunk 4 above)
Comment (1technophile) [reply tests/protocols/iwds07.h]: done

**Comment by CurlyMoo** [comment]
Still some fixing to do when looking at the travis output.

**Comment by 1technophile** [comment]
Yes I have seen that. it seems that I need to change the message creation method.

File: libs/pilight/protocols/433.92/iwds07.c [hunk 5]
```diff
@@ -36,18 +36,24 @@ static int validate(void) {
...
-	iwds07->message=json_mkobject();
-	json_append_member(iwds07->message, "unit", json_mknumber(unit, 0));
-	json_append_member(iwds07->message, "battery", json_mknumber(battery, 0));
-	if(state == 1) {
-		json_append_member(iwds07->message, "state", json_mkstring("closed"));
-	} else {
-		json_append_member(iwds07->message, "state", json_mkstring("opened"));
+static void createMessage(int unit, int battery, int state, char **message) {
+	int x = 0;
+	x = snprintf((*message), 255, "{\"unit\":%d,", unit);
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Please check your alignment and code style.

Comment (1technophile) [reply libs/pilight/protocols/433.92/iwds07.c]: done

**Comment by CurlyMoo** [comment]
Still some compile errors.

**Comment by 1technophile** [comment]
I need to setup a complete dev environment so as to be more efficient. Do you have some recommendations or some guides to point me ?

**Comment by CurlyMoo** [comment]
If you don't have a dev environment, how are you currently testing?

**Comment by 1technophile** [comment]
What I miss currently is unittest running localy to avoid multiple iterations on the CI.

**Comment by CurlyMoo** [comment]
I don't know what you mean.

**Comment by 1technophile** [comment]
It took me some time but I'm close to have a proper environment now with a raspberry pi3 and raspbian installed, in the target of doing proper test for WDS07 and also add more protocols from the different devices I have. 
I was using an ESP8266 previously with ESPilight for info for the tests.

On my raspberry I'm able to compile with `setup.sh` and run the master branch.
Nevertheless with the `rewrite `branch I'm able to compile `setup.sh` but can't run the different pilight executables, indeed there is no pilight files into `/usr/local/sbin/`
When I look into [CMakeLists.txt](https://github.com/pilight/pilight/blob/master/CMakeLists.txt) from master it contains references to the expected files but the [equivalent ](https://github.com/pilight/pilight/blob/rewrite/CMakeLists.txt)from rewrite doesn't have these references.

Am I missing something to do my tests with the rewrite branch?

**Comment by CurlyMoo** [comment]
The rewrite branch can only be used for unittesting. It should be compiled using:
```
cmake -DPILIGHT_UNITTEST=1 \
        -DCOVERALLS=ON \
        -DCMAKE_BUILD_TYPE=Debug ..
make
```
That creates a `pilight-unittest` executable that you execute to run all unittests.

**Comment by 1technophile** [comment]
Thanks for the information.

For the implementation I'm taking the Kerui D026 sensor example.
I'm trying to understand the difference between the [staging ](https://github.com/pilight/pilight/blob/b8d61b4a06322c816098626d0a49f73c1f2c421b/libs/pilight/protocols/433.92/kerui_d026.c#L41)and the [rewrite ](https://github.com/pilight/pilight/blob/rewrite/libs/pilight/protocols/433.92/kerui_d026.c#L41) code.
And I don't catch the fact that the rewrite branch test a different code than the one implemented in staging.
Do I need to replicate these differences into IWDS07 integration to accomodate unit testing?

**Comment by CurlyMoo** [comment]
> Do I need to replicate these differences into IWDS07 integration to accomodate unit testing?

Yes, the reason for that is that i merge changing more slowly that i development them. If i would merge the full rewrite at once, finding issues would be a pain in the ass due to the massive amount of changes.

**Comment by 1technophile** [comment]
Hello,

Could you indicate me the best way to realign my rewrite branch to yours (a squash, a rebase, or other method)?
Or to close the PR refork your code and rebegin grom it?

**Comment by CurlyMoo** [comment]
>  I would like to avoid having all the alignement commits to appears in the PR.

I wouldn't bother about that.

**Comment by 1technophile** [comment]
I'm getting:
```
[ test_lua_async_event_missing_parameters          ]
[ test_lua_async_event                             ]
The command "./ci/travis-script.sh" exited with 134.
```
How could I go deeper with this error?

**Comment by CurlyMoo** [comment]
Can you try again after syncing with the latest rewrite code?

File: libs/pilight/protocols/433.92/iwds07.c [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+}
+
+static void createMessage(int unit, int battery, int state, char **message)
+{
+	int x = 0;
+	x = snprintf((*message), 255, "{\"unit\":%d,", unit);
+	logprintf(LOG_ERR, "create message TEST %d", unit);
+	if (state == 0)
+	{
+		x += snprintf(&(*message)[x], 255 - x, "\"state\":\"closed\"");
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Missing an ending comma.

File: libs/pilight/protocols/433.92/iwds07.c [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	else
+	{
+		x += snprintf(&(*message)[x], 255 - x, "\"state\":\"opened\"");
```
Comment (CurlyMoo) [inline libs/pilight/protocols/433.92/iwds07.c]: Here as well.

File: tests/protocols/iwds07.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+  This Source Code Form is subject to the terms of the Mozilla Public
+  License, v. 2.0. If a copy of the MPL was not distributed with this
+  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+*/
+
+static const int iwds07_nrtests = 8;
+static const struct raw_t iwds07_tests[] = {
+	{
+		"1230 410 410 1230 1640 410 410 1230 1640 410 1640 410 1640 410 410 1230 1230 410 410 1230 410 1230 410 1230 410 1230 410 1230 1230 410 1230 410 410 1230 410 1230 410 1230 410 1230 1230 410 410 1230 1230 410 410 1230 410 13940",
+		"{\"unit\":49525,\"battery\":1,\"state\":\"opened\"}",
```
Comment (CurlyMoo) [inline tests/protocols/iwds07.h]: In your messages the state is output first, than the battery.

**Comment by CurlyMoo** [comment]
I would advice you to run the `test_protocols_433` first to make that work.

So inside your build folder:
```
make -j4 && ./pilight-unittest test_protocols_433
[ test_protocols_433_alecto_ws1700                 ]
[ test_protocols_433_alecto_wx500                  ]
[ test_protocols_433_arctech_contact               ]
[ test_protocols_433_GS-iwds07                     ]
F

There was 1 failure:
1) test_protocols_433: /home/pilight/source/pr/tests/protocols_433.c:111: expected <{"unit":52781,"battery":1,"state":"opened"}> but was <{"unit":52781,"state":"closed","battery":1}>
```

Which clearly show the errors you made.

**Comment by 1technophile** [comment]
Thanks for the feedback, indeed state and battery are not in the good order in the code regarding unittests examples.
Regarding the unit_test unfortunately I'm unable to create the `pilight-unittest` executable in my own environment.
Here is what I'm doing on a raspberry pi3 with stretch

```
sudo su
apt-get install build-essential cmake git dialog libwiringx libwiringx-dev libpcap0.8-dev libmbedtls-dev liblua5.2-dev libluajit-5.1-dev
mkdir rewrite-curly1207
cd rewrite-curly1207
git clone --depth 5 -b rewrite https://www.github.com/pilight/pilight.git
cd pilight
cmake -DPILIGHT_UNITTEST=1 \
        -DCOVERALLS=ON \
        -DCMAKE_BUILD_TYPE=Debug .
make
```

All steps are OK until the last command which after 98% completion give me:
```
[ 98%] Building C object CMakeFiles/pilight-unittest.dir/tests/webserver_rest.c.o
[ 98%] Linking C executable pilight-unittest
libpilight.so: undefined reference to `assert'
collect2: error: ld returned 1 exit status
CMakeFiles/pilight-unittest.dir/build.make:2075: recipe for target 'pilight-unittest' failed
make[2]: *** [pilight-unittest] Error 1
CMakeFiles/Makefile2:142: recipe for target 'CMakeFiles/pilight-unittest.dir/all' failed
make[1]: *** [CMakeFiles/pilight-unittest.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by CurlyMoo** [comment]
Simply add:
```
#include <assert.h>
```
At the top of that file.

**Comment by 1technophile** [comment]
Could you precise the file name please?

**Comment by CurlyMoo** [comment]
It already says in the error:
```
tests/webserver_rest.c
```

**Comment by 1technophile** [comment]
ok my bad

**Comment by 1technophile** [comment]
The include is missing into relay.c, dht11.c and dht22.c
Do you want me to submit another PR for that?

**Comment by CurlyMoo** [comment]
Compilation errors should be fixed.

**Comment by 1technophile** [comment]
Unit tests of 433mhz protocols are ok on my local environment:
```
/home/pi/rewrite1307/pilight/build# make -j4 && ./pilight-unittest test_protocols_433
[  0%] Built target DSO
[  1%] Built target SERIAL
[  1%] Built target GPIO
[  1%] Built target dso_shared
[  1%] Built target gpio_shared
[  2%] Built target serial_shared
[ 70%] Built target pilight
[ 70%] Built target pilight_static
[ 71%] Built target pilight_shared
[100%] Built target pilight-unittest
[ test_protocols_433_alecto_ws1700                 ]
[ test_protocols_433_alecto_wx500                  ]
[ test_protocols_433_arctech_contact               ]
[ test_protocols_433_GS-iwds07                     ]
[ test_protocols_433_arctech_dimmer                ]
[ test_protocols_433_arctech_dusk                  ]
[ test_protocols_433_livolo_switch                 ]
[ test_protocols_433_kerui_D026                    ]
[ test_protocols_433_tfa2017                       ]
.

OK (1 test)
```

**Comment by CurlyMoo** [comment]
So, satisfied?

**Comment by 1technophile** [comment]
Yes !  Thanks for the help.
How do you want me to proceed for the staging modification ?

**Comment by CurlyMoo** [comment]
> How do you want me to proceed for the staging modification ?

Yes

**Comment by 1technophile** [comment]
If I close #437 , rebegin from your staging branch and submit a new PR on staging branch, is it ok for you?

**Comment by CurlyMoo** [comment]
No, PR's should never be closed but should be fixed.

**Pull request merged**